ecosystem/indexer-grpc/indexer-grpc-table-info/src/table_info_service.rs

```diff
@@ -294,6 +373,117 @@ impl TableInfoService {
         Ok(())
     }
 
+    async fn snapshot_indexer_async_v2(
+        context: Arc<ApiContext>,
+        indexer_async_v2: Arc<IndexerAsyncV2>,
+        epoch: u64,
+    ) -> anyhow::Result<()> {
+        let chain_id = context.chain_id().id();
+        // temporary path to store the snapshot
+        let snapshot_dir = context
+            .node_config
+            .get_data_dir()
+            .join(format!("chain_{}_epoch_{}", chain_id, epoch));
+        // rocksdb will create a checkpoint to take a snapshot of full db and then save it to snapshot_path
+        indexer_async_v2
+            .create_checkpoint(&snapshot_dir)
+            .context(format!("DB checkpoint failed at epoch {}", epoch))?;
+
+        Ok(())
+    }
+
+    /// Uploads the snapshot to GCS if found.
+    /// 1. If current epoch is backuped, it will skip the backup.
+    /// 2. If the chain id in the backup metadata does not match with the current network, it will panic.
+    /// Not thread safe.
+    /// TODO(larry): improve the error handling.
+    async fn backup_snapshot_if_present(
+        context: Arc<ApiContext>,
+        backup_restore_operator: Arc<GcsBackupRestoreOperator>,
+    ) {
+        let target_snapshot_directory_prefix = format!("chain_{}_epoch_", context.chain_id().id());
+        // Scan the data directory to find the latest epoch to upload.
+        let mut latest_epoch = None;
```

Comment:
We probably should try to find all epochs that haven't been uploaded.

Reply:
done.